### PR TITLE
GC stress fix for reverse pinvoke stubs

### LINKED ISSUE #13720: Test failed: JIT\Performance\CodeQuality\BenchmarksGame\pidigits\pidigits-3\pidigits-3.cmd
**Job:**
[coreclr-gcstress-extra #20191102.1](https://dev.azure.com/dnceng/public/_build/results?buildId=413324)

**Detail:**
https://helix.dot.net/api/2019-06-17/jobs/684cbdca-f557-48d4-8a7d-b9d9c506a4cf/workitems/JIT/console

**OS & Arch:**
windows x64

**Mode:**
COMPlus_TieredCompilation=0
COMPlus_GCStress=0xC
COMPlus_JitStress=2

**Note:**
This failure **repros** on our windows x64 machine.

**Log:**
```
JIT\Performance\CodeQuality\BenchmarksGame\pidigits\pidigits-3\pidigits-3.cmd [FAIL]
      
      Assert failure(PID 5464 [0x00001558], Thread: 2764 [0x0acc]): SanityCheck()
      
      CORECLR! MethodTable::Validate + 0x3F (0x00007ffa`13215013)
      CORECLR! Object::ValidateInner + 0x127 (0x00007ffa`13136f77)
      CORECLR! Object::Validate + 0x131 (0x00007ffa`13136e11)
      CORECLR! WKS::GCHeap::Promote + 0x9F (0x00007ffa`1356afcf)
      CORECLR! PromoteCarefully + 0xAB (0x00007ffa`1316225b)
      CORECLR! GcEnumObject + 0x61 (0x00007ffa`13476c71)
      CORECLR! GcInfoDecoder::ReportSlotToGC + 0xCD (0x00007ffa`1353e175)
      CORECLR! GcInfoDecoder::EnumerateLiveSlots + 0xCA1 (0x00007ffa`1353d399)
      CORECLR! EECodeManager::EnumGcRefs + 0x39B (0x00007ffa`131a9c6b)
      CORECLR! GcStackCrawlCallBack + 0x2CD (0x00007ffa`1347711d)
          File: F:\workspace\_work\1\s\src\vm\methodtable.cpp Line: 9907
          Image: C:\dotnetbuild\work\684cbdca-f557-48d4-8a7d-b9d9c506a4cf\Payload\CoreRun.exe
      
      
      Return code:      1
      Raw output file:      C:\dotnetbuild\work\684cbdca-f557-48d4-8a7d-b9d9c506a4cf\Work\f0773ebb-06b1-46fb-8104-1bbf4458304f\Exec\JIT\Performance\Reports\JIT.Performance\CodeQuality\BenchmarksGame\pidigits\pidigits-3\pidigits-3.output.txt
      Raw output:
      BEGIN EXECUTION
       "C:\dotnetbuild\work\684cbdca-f557-48d4-8a7d-b9d9c506a4cf\Payload\corerun.exe" pidigits-3.dll 
      Expected: 100
      Actual: -1073740286
      END EXECUTION - FAILED
      FAILED
      Test Harness Exitcode is : 1
```

category:testing
theme:gc-stress
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Reverse PInvoke stubs change the thread to preemptive mode before
they exit. Often the epilog immediately follows this helper call,
but in some cases there may be a region with a few instructions
before we reach the epilog.

If there is a normal GC triggered while the method is in this region
then the runtime will not report GC references from the stub frame.

If the stub method is fully interruptible we will put GC coverage
interrupts in this region when doing GC stress. When these trigger a GC
then the runtime will report GC references from the stub frame.

So if while executing in this region we happen to do a normal GC and then
a stress invoked GC, the stress mode GC can find invalid GC references
in the frame, as the objects referenced could have been collected/moved
by the normal GC that happened just before.

There are a couple of avenues for a fix, but since this is a stress mode
only problem, I've modified GC stress avoid invoking a GC when the thread
is already in preemptive mode and the method is a reverse PInvoke stub.

Fixes #13720.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @jkotas @jkoritzinsky 

More notes / logs / etc in the linked issue.

File: src/coreclr/src/vm/gccover.cpp [hunk 1]
```diff
@@ -1411,6 +1411,32 @@ BOOL OnGcCoverageInterrupt(PCONTEXT regs)
...
+    //
+    // A PInvoke stub should should properly report GC refs if we
+    // trigger GC here. But a reverse PInvoke stub may over-report
+    // leading to spurious failures, as we would not normally report
+    // anything for this method.
+    //
+    // So, suppress GC for the reverse PInvoke stubs.
+    if (!pThread->PreemptiveGCDisabled())
+    {
+        _ASSERTE(pMD->IsILStub());
```
Comment (jkotas) [inline src/coreclr/src/vm/gccover.cpp:1426]:
```suggestion
        _ASSERTE(pMD->HasUnmanagedCallersOnlyAttribute());
```
The method does not have to be IL stub.

Comment (jkotas) [reply src/coreclr/src/vm/gccover.cpp:1426]: Scratch that. It can be any method. PInvoke stub with an inlined frame can appear in any method.

Comment (AndyAyersMS) [reply src/coreclr/src/vm/gccover.cpp:1426]: Good point. That was just some extra paranoia.

**Review by jkotas — APPROVED** [review]
LGTM. Thanks!

**Comment by AndyAyersMS** [comment]
Added some GC stress runs.

**Comment by AndyAyersMS** [comment]
GC stress failures are known issues (#37238, #33948). So no new failures.

**Pull request merged**